crates/sui-core/src/epoch/consensus_store_pruner.rs

```diff
@@ -0,0 +1,253 @@
+                    }
+                    result = rx_remove.recv() => {
+                        if result.is_none() {
+                            info!("Closing consensus store pruner");
+                            break;
+                        }
+                        latest_epoch = result.unwrap();
+                        Self::prune_old_epoch_data(base_path.clone(), latest_epoch, epoch_retention).await;
+                    }
+                }
+            }
+        });
+    }
+
+    /// This method will remove all epoch data stores and directories that are older than the current epoch minus the epoch retention. The method ensures
+    /// that always the `current_epoch` data is retained.
+    pub async fn prune(&self, current_epoch: Epoch) {
+        let result = self.tx_remove.send(current_epoch).await;
+        if result.is_err() {
+            error!(
+                "Error sending message to data removal task for epoch {:?}",
+                current_epoch,
+            );
+        }
+    }
+
+    pub(crate) async fn prune_old_epoch_data(
+        storage_base_path: PathBuf,
+        current_epoch: Epoch,
+        epoch_retention: u64,
+    ) {
+        let drop_boundary = current_epoch.saturating_sub(epoch_retention);
+
+        info!(
+            "Consensus store prunning for current epoch {}. Will remove epochs < {:?}",
+            current_epoch, drop_boundary
+        );
+
+        // Get all the epoch stores in the base path directory
+        let files = match fs::read_dir(storage_base_path) {
+            Ok(f) => f,
+            Err(e) => {
+                error!(
+                    "Can not read the files in the storage path directory for epoch cleanup: {:?}",
+                    e
+                );
+                return;
+            }
+        };
+
+        // Look for any that are less than the drop boundary and drop
+        for file_res in files {
+            let f = match file_res {
+                Ok(f) => f,
+                Err(e) => {
+                    error!(
+                        "Error while cleaning up storage of previous epochs: {:?}",
+                        e
+                    );
+                    continue;
+                }
+            };
+
+            let name = f.file_name();
+            let file_epoch_string = match name.to_str() {
+                Some(f) => f,
+                None => continue,
+            };
+
+            let file_epoch = match file_epoch_string.to_owned().parse::<u64>() {
+                Ok(f) => f,
+                Err(e) => {
+                    error!(
+                        "Could not parse file \"{file_epoch_string}\" in storage path into epoch for cleanup: {:?}",
+                        e
+                    );
+                    continue;
+                }
+            };
+
+            if file_epoch < drop_boundary {
+                if let Err(e) = safe_drop_db(f.path()) {
```

Comment:
If the directory is corrupted, does rocksdb clean up the dir or will repeatedly fail? Maybe after epoch change when retrying, we should still use the original method to delete the directories?

Reply:
Good point. Well on the unit tests I don't even create a proper db and it successfully removes the dirs. But I guess what I can do is fallback to previous approach if the safe one fails? It kind of negates the point of safe delete.

Reply:
If this works in unit tests on dir with any content, it should be fine. I have seen validators / fullnodes with unpruned epoch dbs, and I wondered if that was because of the use of `safe_drop_db()`. But probably it is due to something else.

Reply:
oh ok didn't know that was the case. It would be good to understand under what circumstances this could happen. For consensus db to be honest there shouldn't be many reasons to fail, apart from having some pointer still keeping the db open - which would mostly mean something wrong on the consensus shutdown approach. Since the previous approach served as well so far let me do the fallback to be on the safe side here - we can treat the `safe_drop_db` as the best effort/intention.